AppleALC/ALCUserClientProvider/ALCUserClientProvider.cpp

```diff
@@ -0,0 +1,81 @@
+//
+//  ALCUserClientProvider.cpp
+//  AppleALC
+//
+//  Created by Nick on 10/14/20.
+//  Copyright © 2020 vit9696. All rights reserved.
+//
+
+#include "ALCUserClientProvider.hpp"
+
+OSDefineMetaClassAndStructors(ALCUserClientProvider, IOService);
+
+bool ALCUserClientProvider::start(IOService* provider) {
+	if (!super::start(provider))
+		return false;
+	
+	mHDACodecDevice = IOService::waitForMatchingService(IOService::nameMatching(kIOHDACodecDevice), 3000);	// Wait for 3s
+	
+	if (!mHDACodecDevice)
+	{
+		DBGLOG("client", "HDACodecDevice not found");
+		stop(provider);
+		return false;
+	}
+	
+	// Only attach to internal codec
+	auto codecAddress = OSDynamicCast(OSNumber, mHDACodecDevice->getProperty(kIOHDACodecAddress));
+	if (!codecAddress) {
+		DBGLOG("client", "IOHDACodec address is non-existent");
+		stop(provider);
+		return false;
+	}
+	
+	// Check if attached to correct IOHDACodecDevice
+	if (codecAddress->unsigned8BitValue() != 0) {
+		DBGLOG("client", "IOHDACodec address has a non zero value");
+		stop(provider);
+		return false;
+	}
+	
+	// We are ready for verbs
+	SYSLOG("client", "ALCUserClient is ready for hda-verbs");
+	mHDACodecDevice->setProperty("ReadyForALCVerbs", kOSBooleanTrue);
+	readyForVerbs = true;
+	
+	// Publish the service
+	registerService();
+	
+	return true;
+}
+
+void ALCUserClientProvider::stop(IOService* provider) {
+	super::stop(provider);
+	
+	OSSafeReleaseNULL(mHDACodecDevice);
+}
+
+bool ALCUserClientProvider::init(OSDictionary* dictionary) {
+	if (!super::init(dictionary)) {
```

Comment:
Are you sure you need to override these (init, free) if you do not change the behaviour?

Reply:
Init is required as for some reason if `super::init()` fails, we stop the driver here. Stop is required only in `ALCUserClientProvider` to release the `mHDACodecDevice` instance. Although, we can remove it as user will unlikely ever unload the kext during runtime but having it there won't hurt. What do you say?

Reply:
Well, will not the default implementation, which is essentially super::init do the same?

Reply:
Resolves.